substratest/assets.py

```diff
@@ -168,15 +140,73 @@ def load(cls, d):
             raise errors.TError(f"cannot parse asset `{d}`") from e
 
 
-class _Asset(_DataclassLoader, abc.ABC):
+class _InternalStruct(pydantic.BaseModel, _DataclassLoader, abc.ABC):
+    """Internal nested structure"""
+
+
+class _Asset(_InternalStruct, abc.ABC):
     """Represents assets stored in the Substra framework.
 
     Convert a dict with camel case fields to a Dataclass.
     """
 
 
-@dataclasses.dataclass(frozen=True)
-class Permission(_DataclassLoader):
+class _BaseFutureAsset(_Asset):
+    _future_cls = None
+
+    def attach(self, session):
+        """Attach session to asset."""
+        # XXX because Pydantic doesn't support private fields, we have to use
+        # __getattribute__ and __setattr__ (https://github.com/samuelcolvin/pydantic/issues/655)
+        object.__setattr__(self, '__session', session)
+        return self
+
+    @property
+    def _session(self):
+        __session = object.__getattribute__(self, '__session')
```

Comment:
Are you sure `__session = object.__getattribute__(self, '__session')` will return None / False if `attach` has not been called?

Reply:
Hmm, it actually returns this:
`E       AttributeError: 'Traintuple' object has no attribute '__session'`

Reply:
So it means that the condition inside the next if will always be False.

In this case you will have to use a try/catch syntax.